pkg/cache/node.go

```diff
@@ -0,0 +1,134 @@
+	"github.com/cloudera/yunikorn-core/pkg/log"
+	"github.com/cloudera/yunikorn-k8shim/pkg/common"
+	"github.com/cloudera/yunikorn-k8shim/pkg/common/events"
+	"github.com/cloudera/yunikorn-k8shim/pkg/conf"
+	"github.com/cloudera/yunikorn-scheduler-interface/lib/go/si"
+	"github.com/looplab/fsm"
+	"go.uber.org/zap"
+	"sync"
+)
+
+// stores info about what scheduler cares about a node
+type SchedulerNode struct {
+	name      string
+	uid       string
+	capacity  *si.Resource
+	existingAllocations []*si.Allocation
+	schedulerApi api.SchedulerApi
+	fsm       *fsm.FSM
+	lock      *sync.RWMutex
+}
+
+func newSchedulerNode(nodeName string, nodeUid string,
+	nodeResource *si.Resource, schedulerApi api.SchedulerApi) *SchedulerNode {
+	schedulerNode := &SchedulerNode{
+		name: nodeName,
+		uid:  nodeUid,
+		capacity: nodeResource,
+		schedulerApi: schedulerApi,
+		lock: &sync.RWMutex{},
+	}
+	schedulerNode.initFSM()
+	return schedulerNode
+}
+
+func (n *SchedulerNode) initFSM() {
+	var states = events.States().Node
+	n.fsm = fsm.NewFSM(
+		states.New,
+		fsm.Events{
+			{Name: string(events.RecoverNode),
+				Src: []string{states.New},
+				Dst: states.Recovering,
+			},
+			{Name: string(events.NodeAccepted),
+				Src: []string{states.Recovering},
+				Dst: states.Healthy,
+			},
+			{Name: string(events.NodeRejected),
+				Src: []string{states.New, states.Recovering},
+				Dst: states.Rejected,
+			},
+		},
+		fsm.Callbacks{
+			string(states.Recovering): n.handleNodeRecovery,
+		})
+}
+
+func (n *SchedulerNode) addExistingAllocation(allocation *si.Allocation) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.existingAllocations = append(n.existingAllocations, allocation)
+}
+
+func (n *SchedulerNode) getNodeState() string {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+	return n.fsm.Current()
+}
+
+func (n *SchedulerNode) handleNodeRecovery(event *fsm.Event) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	log.Logger.Info("node recovering",
+		zap.String("nodeId", n.name))
+
+	request := &si.UpdateRequest{
+		Asks:     nil,
+		Releases: nil,
+		NewSchedulableNodes: []*si.NewNodeInfo{
+			{
+				NodeId:              n.name,
+				SchedulableResource: n.capacity,
+				Attributes: map[string]string{
+					common.DefaultNodeAttributeHostNameKey: n.name,
+					common.DefaultNodeAttributeRackNameKey: common.DefaultRackName,
+				},
+				ExistingAllocations: n.existingAllocations,
+			},
+		},
+		RmId: conf.GetSchedulerConf().ClusterId,
+	}
+
+	// send request to scheduler-core
+	if err := n.schedulerApi.Update(request); err != nil {
+		log.Logger.Error("failed to send request",
+			zap.Any("request", request))
+	}
+}
+
+func (n *SchedulerNode) handle(ev events.SchedulerNodeEvent) error {
+	log.Logger.Debug("scheduler node state transition",
+		zap.String("nodeId", ev.GetNodeId()),
+		zap.String("preState", n.fsm.Current()),
+		zap.String("pendingEvent", string(ev.GetEvent())))
+	if err := n.fsm.Event(string(ev.GetEvent()), ev.GetArgs()...); err != nil {
+		return err
```

Comment:
fsm can return an error if there is a transition that does not change the state. Healthy to Healthy i.e. err != nil does not mean the transition really failed:
```
	// handle the same state transition not nil error (limit of fsm).
	if err.Error() == "no transition" {
		return nil
	}
```